# Fragmentomics Analysis
This folder contains scripts and workflows for analyzing fragment length distributions from cell-free DNA (cfDNA) sequencing data for both Nanopore and Illumina platforms.

## Contents

Collecting fragment length for:

**NANOPORE**
1. **bam-nanopore-to-nanoplot.sh:** Filters sorted and aligned BAM files to retain only mapped reads with a mapping quality (MAPQ) â‰¥ 20, re-indexes the filtered BAM files, and runs NanoPlot to collect fragment length statistics and plots.
3. **change_format_nanoplot.py:** Formats the .tsv.gz output from NanoPlot into a cleaner and reduced size format compatible with downstream analyses.

**ILLUMINA**
1. **bam-to-picard.sh:** Filters BAM files (keeping mapped reads with MAPQ â‰¥ 20), then uses Picard CollectInsertSizeMetrics to extract insert size metrics and generate insert size histograms.

## Usage

* **bam-nanopore-to-nanoplot.sh:**
  **Input:**
  * Folder containing BAM files (aligned and sorted) with their corresponding index (.bai).

  **Output:**
  * Filtered BAM files.
  * Plots generated by Nanoplot read length distributions, quality plots, etc).
  * .tsv.gz files containing fragment length information.

  **Requirements:**
  * NanoPlot
  * samtools
  * GNU parallel

  **Commands:**
  
    ```bash
    ./bam-nanopore-to-nanoplot.sh <folder_bam> <output_folder>
    ```


    **Installation:**
     ```bash
    conda install -c bioconda nanoplot samtools parallel
     ```

----
* **change_format_nanoplot.py:**
  **Input:**
  * .tsv.gz file generated by NanoPlot.

  **Output:**
  * Reformatted fragment length file.

  **Command:**

  Change the input and output folder variable inside the script. 
  ```bash
    python3 change_format_nanoplot.py
  ```

----
* **bam-to-picard.sh:**

ðŸ”¸ bam-to-picard.sh
Inputs:

Folder containing BAM files (aligned and sorted).

Outputs:

Filtered BAM files.

*.txt files with insert size metrics per sample.

*.pdf histograms of fragment length distributions.

Requirements:

Picard

samtools

Command:

bash
Copy
Edit
./bam-to-picard.sh <folder_bam> <output_folder>
Installation Example:

bash
Copy
Edit
conda install -c bioconda picard samtools
# or
conda install -c bioconda/label/cf201901 picar

    **Requirements:**
    * samtools
    * picard 

  **Input:**
  * Folder with BAM files.
<input_tsv.gz> <output_file>
  **Output:**
  * Folder with filtered BAMs.
  * *.txt (insert sizes per sample).
  * *.pdf (histogram of fragment distributions per sample).

  **Command:**
  
  ```sh
  ./bam-to-picard.sh <folder_bam> <output_folder>
  ```
  
  **Installation:**
  
  ```sh
  conda install bioconda::picard
  # or
  conda install bioconda/label/cf201901::picard

  ```

